Fixed bug when used "\\" at the end of string

vsile opened the pull request:
```
{
  "metrics": [
    {
        "path": "dpc\\root\\GRs\\PK\\gr047_mirror",
        "value": "0"
    },
    {
        "path": "dpc\\root\\GR\\PK\\",
        "value": "1"
    },
    {
        "path": "dpc\\root\\GRs\\PK\\gr048_mirror",
        "value": "1"
    }
  ]
}
```
The next command returns an error because backslash before quote.
```
$ ./jsonpath --file=test_export.json --path='$.metrics[*].path+' --keys
```
In this case backslash is escaped and there is no syntax error.
So we need to check the character before previous.

vsile committed:
Fixed bug when used "\\" at the end of string

```
{
  "metrics": [
    {
        "path": "dpc\\root\\GRs\\PK\\gr047_mirror",
        "value": "0"
    },
    {
        "path": "dpc\\root\\GR\\PK\\",
        "value": "1"
    },
    {
        "path": "dpc\\root\\GRs\\PK\\gr048_mirror",
        "value": "1"
    }
  ]
}
```
The next command returns an error because backslash before quote.
```
$ ./jsonpath --file=test_export.json --path='$.metrics[*].path+' --keys
```
In this case backslash is escaped and there is no syntax error.
So we need to consider the character before previous.

kawamuray commented:
Thanks for the patch :)
Curious what happens if the input string is like `"dpc\\root\\GR\\PK\\\""`. In this case  the parsed string is supposed to be `dpc\root\GR\PK\"` but the patched code produces a string without trailing dquote I think?

vsile committed:
Update lexer_reader.go

The next cases are taken into consideration
1. Escaped quote before closing double qoute
```
"path": "dpc\\root\\GR\\PK\""
```
2. Three backslashes before quote.
```
"path": "dpc\\root\\GR\\PK\\\""
```

vsile commented:
You're right.
Also I should note that even in previous original code we can't use escaped double quote before closing double quote
```
{
  "metrics": [
    {
        "path": "dpc\\root\\GRs\\PK\\gr047_mirror",
        "value": "0"
    },
    {
        "path": "\"Some text\"",
        "value": "1"
    },
    {
        "path": "dpc\\root\\GRs\\PK\\gr048_mirror",
        "value": "1"
    }
  ]
}
```
Run jsonpath and get error
```
$ ./jsonpath --file=test_export.json --path='$.metrics[*].path+' --keys
"metrics"	0	"path"	"dpc\\root\\GRs\\PK\\gr047_mirror"
Unexpected character after json value token: U+0076 'v' at byte index 155
```

In the current patch we have solve this issue and take in consideration the next cases:
```"path": "dpc\\root\\GR\\PK\\"```
```"path": "dpc\\root\\GR\\PK\\\""```

kawamuray commented:
Thanks for the follow-up. 

I think this implementation still fails for input like `PK\\\\"`. In this case, when `curByte` reaches to the last dquote, all of `previous`, `byteBeforePrevious` and `secondByteBeforePrevious` points to a backslash so it doesn't break out of the loop, but the intention of the input string is like 1st backslash escapes the 2nd, and 3rd backslash escapes the 4th, so the dquote isn't escaped.

I'm not fully sure if I understand the original implementation correctly, since this repository was a fork of another original (but the original repository has gone so now this repo looks like the original), but I think correct backslash escaping needs to be implemented by "looking forward" rather than "looking back"? something like below:

```
for {
  curByte = ...
  l.lexeme.WriteByte(curByte)

  if curByte == '\\' {
      // look forward
      next = bufInput.ReadByte();
      if next is not null {
          WriteByte(next);
      }
  } else if curByte == '"' {
     // this dquote is not escaped
     break;
  }
}
```

kawamuray reviewed (changes requested):
Please see the above comment.

vsile committed:
Update lexer_reader.go

Now we count the backslashes before double quote.
If the number of backslashes before double quote is even then break loop.
`"`, `\\"`, `\\\\"`, etc
If odd then it means that double quote were escaped.
`\"`, `\\\"`, `\\\\\"`, etc

vsile commented:
You're right again. I have updated code.

Now we count the backslashes before double quote.
If the number of backslashes before double quote is even then break loop.
`"`, `\\"`, `\\\\"`, etc
If odd then it means that double quote were escaped.
`\"`, `\\\"`, `\\\\\"`, etc

kawamuray reviewed (approved):
Ok, I think that approach works too. LGTM, thanks 👍
